### ISSUE TITLE
IAM vs EC2 bound role doc mismatch

### ISSUE THREAD

**Issue by holtwilkins** [hashicorp/vault#3472]
**Environment:**

* Vault Version: `Vault v0.8.3 ('6b29fb2b7f70ed538ee2b3c057335d706b6d4e36')`

**Expected Behavior:**
I would expect either the `bound_iam_role_arn` documentation in `path-help` to **not** say it's only checked when auth_type is ec2, since this is seemingly false, or that using IAM auth when only this is set would fail.

**Actual Behavior:**
I have the following role set

```
$ vault read auth/aws/role/iam-test
Key                           	Value
---                           	-----
allow_instance_migration      	false
auth_type                     	iam
bound_account_id
bound_ami_id
bound_iam_instance_profile_arn
bound_iam_principal_arn
bound_iam_principal_id
bound_iam_role_arn            	arn:aws:iam::1234567890:role/test-role
bound_region
bound_subnet_id
bound_vpc_id
disallow_reauthentication     	false
inferred_aws_region           	us-east-1
inferred_entity_type          	ec2_instance
max_ttl                       	0
period                        	259200
policies                      	[test-policy]
resolve_aws_unique_ids        	true
role_tag
ttl                           	0
```

This policy uses IAM `auth_type` and only specified `bound_iam_role_arn`.  I'm able to auth on an ec2 instance using the bound role, but when I try on another instance, I get

```
* error validating instance: IAM role ARN "arn:aws:iam::1234567890:role/other-role" does not satisfy the constraint role "iam-test"
```

So, this seems like either a documentation bug on the `bound_iam_role_arn` parameter or an authentication bug. 

This is the section of the doc I'm referring to (run `vault path-help auth/aws/role/asdf`):

```
    bound_iam_role_arn (string)
        If set, defines a constraint on the authenticating EC2 instance
        that it must match the IAM role ARN specified by this parameter.
        The value is prefix-matched (as though it were a glob ending in
        '*').  The configured IAM user or EC2 instance role must be allowed
        to execute the 'iam:GetInstanceProfile' action if this is
        specified. This is only checked when auth_type is
        ec2.
```

But then to add more confusion, this block about the IAM `auth_type` references `bound_iam_role_arn`, but not `bound_iam_principal_arn`:

```
    inferred_entity_type (string)
        When auth_type is iam, the
        AWS entity type to infer from the authenticated principal. The only supported
        value is ec2_instance, which will extract the EC2 instance ID from the
        authenticated role and apply the following restrictions specific to EC2
        instances: bound_ami_id, bound_account_id, bound_iam_role_arn,
        bound_iam_instance_profile_arn, bound_vpc_id, bound_subnet_id. The configured
        EC2 client must be able to find the inferred instance ID in the results, and the
        instance must be running. If unable to determine the EC2 instance ID or unable
        to find the EC2 instance ID among running instances, then authentication will
        fail.
```

So I'm guessing this is mostly a documentation bug, but just want to make sure.

**Comment by joelthompson** [comment]
Whoops, sorry about that, definitely a documentation bug. The behavior of the code was changed during development but the path help just hadn't been updated. Should be fixed by #3474 

> But then to add more confusion, this block about the IAM `auth_type` references `bound_iam_role_arn`, but not `bound_iam_principal_arn`

`bound_iam_principal_arn` is very specific to just the iam `auth_type` whereas `inferred_entity_type` is a way of "importing" EC2-specific binds into the iam `auth_type`.

**Comment by holtwilkins** [comment]
LGTM, thanks!
